core/discovery/storage.go

```diff
@@ -41,14 +41,17 @@ type ProposalFilter interface {
 }
 
 // NewStorage creates new instance of ProposalStorage
-func NewStorage() *ProposalStorage {
+func NewStorage(eventPublisher Publisher) *ProposalStorage {
```

Comment:
I saw Publisher interface in 10 places. There is already pacakge eventbus which contains this interface

```go
package eventbus

// Publisher is responsible for publishing given events.
type Publisher interface {
	Publish(topic string, data interface{})
}
```

I suggest to remove all other interface copies and use this one.

Reply:
Improved, just not sure what it improves. Usually Go encourages to have interfaces on caller side.